drivers/video/hm0360.c

```diff
@@ -0,0 +1,930 @@
+	};
+
+	if (!i2c_is_ready_dt(&config->bus)) {
+		/* I2C device is not ready, return */
+		return -ENODEV;
+	}
+
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(pwdn_gpios)
+	/* Power up camera module */
+	if (config->pwdn.port != NULL) {
+		if (!gpio_is_ready_dt(&config->pwdn)) {
+			return -ENODEV;
+		}
+		ret = gpio_pin_configure_dt(&config->pwdn, GPIO_OUTPUT_ACTIVE);
+		if (ret < 0) {
+			LOG_ERR("Could not clear power down pin: %d", ret);
+			return ret;
+		}
+	}
+#endif
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
+	/* Reset camera module */
+	if (config->reset.port != NULL) {
+		if (!gpio_is_ready_dt(&config->reset)) {
+			return -ENODEV;
+		}
+		ret = gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT_ACTIVE);
+		if (ret < 0) {
+			LOG_ERR("Could not set reset pin: %d", ret);
+			return ret;
+		}
+		/* Reset is active high, has 1ms settling time*/
+		k_msleep(1);
+		gpio_pin_set_dt(&config->reset, 0);
+		k_msleep(1);
+	}
+#endif
+
+	uint32_t pid = 0x0000;
+
+	/*  Read Product ID and check if PID OK  */
+	ret = video_read_cci_reg(&config->bus, HM0360_CCI_ID, &pid);
+	if (ret < 0) {
+		LOG_ERR("Could not read product ID: %d", ret);
+		return ret;
+	}
+
+	if (pid != hm0360_PROD_ID) {
+		LOG_ERR("Incorrect product ID: 0x%02X", pid);
+		return -ENODEV;
+	}
+	LOG_DBG("Product ID is: %x", pid);
+
+	/* Reset camera registers */
+	ret = video_write_cci_reg(&config->bus, HM0360_CCI_RESET, 0x01);
+	if (ret) {
+		LOG_ERR("Could not write reset: %d", ret);
+		return ret;
+	}
+	/* Delay for 1ms. */
+	k_msleep(1);
+
+	/* Write Default registers */
+	ret = video_write_cci_multiregs16(&config->bus, hm0360_default_regs,
+					  ARRAY_SIZE(hm0360_default_regs));
+	if (ret < 0) {
+		LOG_ERR("Failed to set default registers");
+		return ret;
+	}
+
+	ret = hm0360_set_fmt(dev, &fmt);
+	if (ret) {
+		return ret;
+	}
+
+	frmival.numerator = 1;
+	frmival.denominator = 15;
+	ret = hm0360_set_frmival(dev, &frmival);
+	if (ret < 0) {
+		LOG_ERR("ERROR: Unable to set up video format");
+		return ret;
+	}
+
+	/* Initialize controls */
+	return hm0360_init_controls(dev);
+}
+
+static int hm0360_set_stream(const struct device *dev, bool enable, enum video_buf_type type)
+{
+	int ret;
+	const struct hm0360_config *config = dev->config;
+
+	ret = video_write_cci_reg(&config->bus, HM0360_CCI_MODE, HM0360_MODE_STREAMING);
+	if (ret < 0) {
+		LOG_ERR("Could not set streaming");
+		return ret;
+	}
+	k_msleep(4);
+
+	return 0;
+}
+
+static int hm0360_set_ctrl(const struct device *dev, uint32_t id)
+{
+	const struct hm0360_config *config = dev->config;
+	struct hm0360_data *drv_data = dev->data;
+	struct hm0360_ctrls *ctrls = &drv_data->ctrls;
+
+	int ret;
+	uint32_t data;
+
+	switch (id) {
+	case VIDEO_CID_HFLIP:
+		ret = video_read_cci_reg(&config->bus, HM0360_CCI_Orientaton, &data);
+		if (ret < 0) {
+			return ret;
+		}
+		ret = video_write_cci_reg(&config->bus, HM0360_CCI_Orientaton,
+					  HM0360_SET_HMIRROR(data, ctrls->hflip.val));
+		if (ret < 0) {
+			return ret;
+		}
```

Comment:
If you wish you can use `video_modify_cci_reg()` and `FIELD_PREP()` this way:

```c
#define HM0360_ORIENTATION_HMIRROR BIT(0)
```

```c
		ret = video_modify_cci_reg(&config->bus, HM0360_CCI_ORIENTATION,
					FIELD_PREP(HM0360_ORIENTATION_HMIRROR, ctrls->hflip.val),
					HM0360_ORIENTATION_HMIRROR);
		if (ret < 0) {
			return ret;
		}
```

This automates the work you did for `#define HM0360_SET_HMIRROR(r, x)      ((r & 0xFE) | ((x & 1) << 0))` but in a generic way for all drivers.

This mean you only need to define a mask, and the `FIELD_GET()` and `FIELD_PREP()` macros can take care of fitting the value inside the mask position. `GENMASK(5, 3)` is usable as a replacement for `BIT(5) | BIT(4) | BIT(3)` in case you wanted a quick way to generate a `[5:3]` bit mask as seen in datasheets.